Audio Debugging View

keianhzo opened the pull request:
Closes https://github.com/mozilla/hubs/issues/4165
Closes https://github.com/mozilla/hubs/issues/4166

This PR adds:
- An `audio-debugging-system` and a view that that can be enabled from preferences and renders the audio sources elements and it's parameters.
- A `audio-gain-system` that handles all the audio gains and clips audio when attenuation is below certain threshold

netpro2k commented:
Still haven't gotten around to reviewing this but I will try to soon.... 

The one thing I noticed off hand is that I think we should not change any falloff defaults in this PR. For me one of the main goals of the debug view will be to help us figure out what better defaults are. So we should first ship just the debug view (and hopefully some way to easily play with values) and then come up with defaults by getting in a room with a bunch of people and messing around with values till we find something we are happy with.

keianhzo commented:
I think the only one that I've changed is the `avatarConeInnerAngle` that by default is 360 and I've changed it to 180, I thought that one was a must but I can revert. `avatarConeOuterAngle` and `avatarConeOuterGain` are set to the default value. I've just added them to the settings so we can eventually also change them from the scene settings in Spoke.

I wanted to jump right away to the OB blitting issue so I haven't pushed the tweaking panel but I'll add that too.

netpro2k commented on src/utils/audio-normalizer.js:
```diff
@@ -0,0 +1,89 @@
+// Inserts analyser and gain nodes after audio source.
+// Analyses audio source volume and adjusts gain value
+// to make it in a certain range.
+export class AudioNormalizer {
```
Hard to compare the diff here since it was moved from another file. Is this just a direct copy/paste?

keianhzo replied:
Yes, this is just a copy paste of what live inside the avatar-volume-controls component.

netpro2k commented on src/systems/audio-debug-system.js:
```diff
@@ -0,0 +1,323 @@
+  }
+`;
+
+AFRAME.registerSystem("audio-debug", {
```
Awesome job on the visuals here, looks really slick, and this will definitely be super useful for both us tuning default audio settings as well as space designers for tuning their spaces. Bet it was pretty fun to write too :)

keianhzo replied:
Yeah, I'm pretty happy about how is looking, I wanted it to be a elegant, simple and clear and I think it's pretty ok. Definitely fun stuff.

netpro2k commented on src/systems/audio-debug-system.js:
```diff
@@ -0,0 +1,323 @@
+      return;
+    }
+
+    this.sourcePositions.fill(new THREE.Vector3());
```
Same as mentioned elsewhere, avoid creating threejs objects every frame. For these in particular you are actually copying the reference anyway so I don't think you even need to update these every frame, just when a new source is added or removed.

keianhzo replied:
Yep, I've moved all that to the initalization.

netpro2k commented on src/systems/audio-debug-system.js:
```diff
@@ -0,0 +1,323 @@
+
+    this.sourcePositions.fill(new THREE.Vector3());
+    this.sourceOrientations.fill(new THREE.Vector3());
+    this.distanceModels.fill(0);
```
Since we are storing count the shader is going to ignore any values we don't immediately overwrite, so not sure we need to actually clear any of these... Not super important to optimize a debug view anyway but figured I would mention it.

netpro2k commented on src/systems/audio-debug-system.js:
```diff
@@ -0,0 +1,323 @@
+    vUv = position.xz;
+  }
+`;
+const CONE_FRAG = `
```
Just FYI our webpack is tooled out so you can make these external like `import mediaHighlightFrag from "./media-highlight-frag.glsl";` Not a huge deal to inline bu this one is reasonably long so it might not hurt. You would need to make MAX_DEBUG_SOURCES a define instead (ShaderMaterial will handle inserting it for you)

keianhzo replied:
Great, that's much cleaner. Refactored.

netpro2k commented on src/systems/audio-debug-system.js:
```diff
@@ -0,0 +1,323 @@
+    }
+    
+    // Blend
+    gl_FragColor = vec4(background.rgb, background.a * vNormal.y);
```
Why does alpha get multiplied by the y component of the normal?

keianhzo replied:
To grade the alpha based on y normal, basically so we paint on the floor but not on the walls sort to speak.

keianhzo replied:
Actually I've just seen that we are doing it twice as we also do it during the attenuation calculation and that's not really necessary.

netpro2k commented on src/systems/audio-debug-system.js:
```diff
@@ -0,0 +1,323 @@
+    return vec4( rgb, circle);
+  }
+
+  vec4 att(float d, vec4 circle, float maxDistance, float refDistance, float rolloffFactor, int distanceModel) {
```
Fairly minor since this a debug shader and "computers are fast", but we re-caculate attenuation a few times. 

Should maybe be split into:
```glsl
float att(float d,float maxDistance, float refDistance, float rolloffFactor, int distanceModel)
```
and
```glsl
vec4 draw_att(float attenuation, vec4 circle)
```

keianhzo replied:
Good call, updated

netpro2k commented on src/systems/audio-debug-system.js:
```diff
@@ -0,0 +1,323 @@
+      float g = clamp(gain[i], 0.0, 1.0);
+      vec4 gainLayer = circle(center, d, g, 1.0, 0.5, colorGain, startOffset);
+      background = mix(background, gainLayer, gainLayer.a);
+      if (gain[i] > 1.0) {
```
In case its helpful, I was originally going to say that from what I read using `fwidth` inside a conditional is sort of undefined behavior, because the threads that are inactive in the conditional are still sampled but that register would have an arbitrary value. But in this case all threads are going to be executing the conditional in the same way since its only dependent on the uniforms, which are by definition uniform across all threads.

keianhzo replied:
I didn't know that, I've been reading and it makes sense the way modern GPUs work but yeah being uniform dependant should be ok. I'm not sure how optimized this shader is, I guess it can be pushed a little bit more, not sure if the time/perfomance gain makes sense though.

netpro2k replied:
Yeah this is only for debugging anyway so I don't think we need to really worry about perf on it very much at all. Just hadn't heard of `fwidth` before and noticed this while reading about it.

netpro2k reviewed (commented):
This is looking great! I love the debug view and the debug panel will be really helpful for coming up with better defaults. We still have a few more stages of refactoring to clean up audio handling but this is a great first step that at least gives us the target for "source of truth" then we can start reading from. Then we can migrate it to actually being the source of truth and remove all the other places where we define these settings.

keianhzo commented on src/components/audio-params.js:
```diff
@@ -0,0 +1,238 @@
+    position: { type: "vec3", default: { x: 0, y: 0, z: 0 } },
+    orientation: { type: "vec3", default: { x: 0, y: 0, z: 0 } },
+    distanceModel: { default: "inverse", oneOf: ["linear", "inverse", "exponential"] },
+    rolloffFactor: { default: 3 },
```
Right, It seems that the defaults for avatar is 3 and media is 1. audio-target is set to 5. I have reverted to 1 for consistency until we set better default for everything.

keianhzo commented on src/components/audio-params.js:
```diff
@@ -0,0 +1,238 @@
+    distanceModel: { default: "inverse", oneOf: ["linear", "inverse", "exponential"] },
+    rolloffFactor: { default: 3 },
+    refDistance: { default: 1 },
+    maxDistance: { default: 20 },
```
Same as above.

keianhzo commented on src/components/audio-params.js:
```diff
@@ -0,0 +1,238 @@
+    rolloffFactor: { default: 3 },
+    refDistance: { default: 1 },
+    maxDistance: { default: 20 },
+    clippingThreshold: { default: 0.015 },
```
I'll just disable clipping by default to keep the current behavior.

keianhzo commented on src/components/audio-params.js:
```diff
@@ -0,0 +1,238 @@
+
+const MUTE_DELAY_SECS = 1;
+
+AFRAME.registerComponent("audio-params", {
```
Yeah, right now this is still updated on every tick with the original source parameters but next step should be making it the source of truth for every audio component.

keianhzo commented on src/components/audio-params.js:
```diff
@@ -0,0 +1,271 @@
+const MUTE_DELAY_SECS = 1;
+
+AFRAME.registerComponent("audio-params", {
+  dependencies: ["media-views", "avatar-audio-source"],
```
Humm, I thought it just affected the components initialization order. Removed.

keianhzo commented on src/components/audio-params.js:
```diff
@@ -0,0 +1,271 @@
+    } else if (this.el.components["media-video"]) {
+      this.data.sourceType = SourceType.MEDIA_VIDEO;
+      this.data.gain = this.el.components["media-video"].data.volume;
+      this.el.addEventListener("media-volume-changed", this.volumeUpdated.bind(this));
```
Fixed, thanks!

keianhzo commented on src/components/audio-params.js:
```diff
@@ -0,0 +1,271 @@
+  },
+
+  audio() {
+    if (this.data.sourceType === SourceType.AVATAR_RIG) {
```
Not sure if I get this on. Do you refer to the returned "fake" audio object for the avatar rig? The audio object for media and avatar-audio-source can change anytime so we always need the latest version. Same for the avatar rig, the audio settings may change.

keianhzo commented on src/components/audio-params.js:
```diff
@@ -0,0 +1,271 @@
+    } else {
+      // We one only enable the Normalizer for avatar-audio-source components
+      if (this.data.sourceType === SourceType.AVATAR_AUDIO_SOURCE) {
+        this.enableNormalizer();
```
We are calling `enableNormalizer` only in case it's not already initialized:
```
if (this.normalizer !== null) {
      this.normalizer.apply();
    } else {
      // We one only enable the Normalizer for avatar-audio-source components
      if (this.data.sourceType === SourceType.AVATAR_AUDIO_SOURCE) {
        this.enableNormalizer();
      }
    }
}
```
We can't call it on `init` as the audio element might not have been initialized yet or it might have changed and we would end up with an outdated audio element: https://github.com/mozilla/hubs/blob/565e944f5a2dac5a8b571a031a6308dabbfe370a/src/components/avatar-audio-source.js#L140

keianhzo commented on src/components/audio-params.js:
```diff
@@ -0,0 +1,271 @@
+      };
+    } else if (this.data.sourceType === SourceType.MEDIA_VIDEO) {
+      const audio = this.el.getObject3D("sound");
+      return audio instanceof THREE.PositionalAudio ? audio : null;
```
Neat! fixed

keianhzo commented on src/components/audio-params.js:
```diff
@@ -0,0 +1,271 @@
+      const avatarAudioSource = this.el.components["avatar-audio-source"];
+      if (avatarAudioSource) {
+        this.normalizer = new AudioNormalizer(audio);
+        avatarAudioSource.el.addEventListener("sound-source-set", () => {
```
Fixed

keianhzo commented on src/components/audio-params.js:
```diff
@@ -0,0 +1,271 @@
+  },
+
+  updateDistances() {
+    const listenerPos = new THREE.Vector3();
```
True, moved to a component variable to create at init only.

netpro2k replied:
For temporary state light this I usually use a closure or a variable with the `tmp` prefix just defined outside of the component so that its not confused as some piece of state we actually care about keeping around over time. Pretty minor though.

keianhzo commented on src/components/audio-params.js:
```diff
@@ -0,0 +1,271 @@
+    }
+  },
+
+  clipGain(gain) {
```
Yep, it's that way just in case we would like to change the clipping gain to make it completely clip or just set it at a fixed really low gain. I guess it doesn't really make much sense...

keianhzo commented on src/components/audio-params.js:
```diff
@@ -0,0 +1,271 @@
+  },
+
+  updateAttenuation() {
+    if (this.data.distanceModel === "linear") {
```
Good call, more elegant indeed. Updated.

keianhzo commented on src/components/audio-params.js:
```diff
@@ -0,0 +1,271 @@
+    } else if (this.data.sourceType === SourceType.AVATAR_AUDIO_SOURCE) {
+      globalVolume = globalVoiceVolume;
+    }
+    const volumeModifier = (globalVolume !== undefined ? globalVolume : 100) / 100;
```
Noted. I'll open an issue with all the things we have been talking about for a second pass.

keianhzo commented on src/components/avatar-audio-source.js:
```diff
@@ -139,14 +131,24 @@ AFRAME.registerComponent("avatar-audio-source", {
     this.destroyAudio();
+  },
+
+  setPositionalAudioProperties(audio, settings) {
```
I agree. I think I moved it at some point and never reverted it to the original state. Updated.

keianhzo commented on src/hub.html:
```diff
@@ -220,7 +220,8 @@
                         <template data-name="Head">
                             <a-entity
-                                avatar-audio-source="rolloffFactor: 2.0;"
+                                avatar-audio-source="rolloffFactor: 3.0;"
```
Fixed.

keianhzo commented on src/react-components/debug-panel/AudioDebugPanel.js:
```diff
@@ -0,0 +1,631 @@
+import React, { useCallback, useState, useRef, useEffect } from "react";
```
Definitely much better, I just used the same for consistency with the previous RTC interface so both can co-live in harmony. Noted for the second pass.

netpro2k replied:
Yeah I don't think we have to do it one way or the other, but for stuff that's strictly debugging we don't particularly care how its implemented so whatever is the quickest and easiest seems good.

keianhzo commented on src/react-components/preferences-screen.js:
```diff
@@ -484,6 +494,18 @@ const preferenceLabels = defineMessages({
+  },
+  enableAudioClipping: {
+    id: "preferences-screen.preference.enable-audio-clipping",
+    defaultMessage: "Enable Audio Clipping"
```
So we removed them from the preferences panel and we don't apply clipping at all unless you are in the audio debugging mode (until we have spoke/blender integration)? Or we default it as disabled?

netpro2k replied:
My argument is even once we decide to use clipping this would not be a user preference but one related to the scene audio properties and falloff settings for audio zones and such. Not sure this makes sense as a user pref.

keianhzo commented on src/systems/audio-gain-system.js:
```diff
@@ -0,0 +1,77 @@
+
+const CLIPPING_GAIN = 0.0001;
+
+AFRAME.registerSystem("audio-gain", {
```
I've rafactored into a Hubs system. Other than the execution order, is there any other reason why a system should be a Hubs system or AFrame system?

netpro2k replied:
That's the primary reason. The other reason just being we are slowly working to factor down the amount of aframe specific code we have.

keianhzo commented:
@netpro2k Thanks a lot for the thorough review, it definitely looks better. I've a few questions but otherwise I think I've addressed all your concerns.

keianhzo commented:
I've also added debug support for `audio-target`.

netpro2k commented on src/components/audio-params.js:
```diff
@@ -0,0 +1,271 @@
+      const worldQuat = new THREE.Quaternion();
+      avatarRigObj.getWorldQuaternion(worldQuat);
+      this.avatarRigOrientation.applyQuaternion(worldQuat);
+      return {
```
This is allocating several new objects per avatar every frame. I think that is going to lead to noticeable GC pressure.

keianhzo replied:
Moved this to a temp var to avoid allocation

keianhzo commented on src/systems/audio-debug.frag:
```diff
@@ -0,0 +1,161 @@
+// Based on: https://www.shadertoy.com/view/Mll3D4#
+
+#define MAX_DEBUG_SOURCES 64
```
oh that's cool, updated!

keianhzo commented:
@netpro2k Updated with the latest change requests.
